cloudwatch-logs-alerts/lambda/src/error-subscriber.ts

```diff
@@ -0,0 +1,176 @@
+import { PublishCommand, SNSClient } from "@aws-sdk/client-sns";
+import { CloudWatchLogsEvent } from "aws-lambda";
+import { gunzipSync } from "node:zlib";
+
+const stage = process.env.STAGE;
+const primaryRegion = process.env.PRIMARY_AWS_REGION;
+const snsTopicEmailArn = process.env.SNS_TOPIC_EMAIL_ARN;
+const snsTopicChatbotArn = process.env.SNS_TOPIC_CHATBOT_ARN;
+
+// https://stackoverflow.com/questions/60796991/is-there-a-way-to-generate-the-aws-console-urls-for-cloudwatch-log-group-filters
+function getLogEventsUrl(
+  logGroupRegion: string,
+  logGroup: string,
+  logStream: string
+) {
+  return `https://console.aws.amazon.com/cloudwatch/home?region=${logGroupRegion}#logEventViewer:group=${logGroup};stream=${logStream}`;
+}
+
+// from lambda@edge cloudwatch log subscription filter events, there's no way to get the region from the event itself,
+// so we parse it from the subscription filter name as the format is in *our* control and does include the region in the name
+function resolveEventRegion(subscriptionFilters: string[] | undefined): string {
+  let region = primaryRegion; // default to primary region
+
+  if (subscriptionFilters?.length) {
+    // Match the region from the subscription filter name, which is defined in the pulumi cloudwatch related definitions
+    // Eg. codesets-EdgeRegion-CloudWatchLogSubFilter-eu-central-1-dev-c1c1724 -> eu-central-1
+    const regexp = new RegExp(
+      `codesets-EdgeRegion-CloudWatchLogSubFilter-(.*)-${stage}-(.*)`
```

Comment:
Tämä regexp toimii nykyisellä resurssien nimeämistavalla vain kun viesti tulee codesets-projektista, koska staattinen etuliite `codesets-`, mutta sepä varmaan ihan ok kun muissa projekteissa ei ole tuota ongelmaa että tarvitsee region selvittää tällä viisiin nimestä.  Jos kuitenkin jättäisi mahdollisuuden että toinenkin projekti käyttäisi tuota patternia logifiltterin nimestä regionin selvittämiseen niin tuon voisi muuttaa esim. näin: `(.*)-EdgeRegion-CloudWatchLogSubFilter-(.*)-${stage}-(.*)` ja sit region-mätsäys kaivettaisiin `region = match[2]`.

Reply:
Tätä juuri pohdiskelin tässä miten parantelisi, koska tämä tosiaan vaan olettaa codesetsia (ja niitä edge regioneita) - muutin tuota skeemaa codesetsin ja cache-updater filtereiden välillä. Myös se että tämä olettaa, että jos matchia ei tule, niin silloin regionin täytyy olla määritelty primary region -> ei myöskään voi varmuudella pitää paikkaansa, jos tähän liitetään muitakin serviiseitä.

Mietin jopa sellasaista, että jos muista sovelluksista exportattais tieto, mihin regioniin mikäkin kuuluu, ja tyyliin kun on saatu subjecti selville, niin sen mukaan mätsätään tuo region tieto.. mutta siinä alkaa olla jo vähän circular deps meininkiä ja pitäis aina tietää ja muistaa exporttaa tuo tieto muista sovelluksista. Harmi kyllä että tuota tietoa ei voi mitenkään järkevästi kaivaa eventistä itestään.